framework/src/problems/FEProblemBase.C

```diff
@@ -3954,6 +3961,72 @@ FEProblemBase::execute(const ExecFlagType & exec_type)
 
   // Return the current flag to None
   setCurrentExecuteOnFlag(EXEC_NONE);
+
+  if (_uo_aux_state_check && !_checking_uo_aux_state)
+  {
+    // we will only check aux variables and postprocessors
+    // checking more reporter data can be added in the future if needed
+    std::unique_ptr<NumericVector<Number>> x = _aux->currentSolution()->clone();
+    DenseVector<Real> pp_values = getReporterData().getAllRealReporterValues();
+
+    // call THIS execute one more time for checking the possible states
+    _checking_uo_aux_state = true;
+    FEProblemBase::execute(exec_type);
+    _checking_uo_aux_state = false;
+
+    *x -= *_aux->currentSolution();
+    if (x->l2_norm() > 1e-8)
+    {
+      const auto & sys = _aux->system();
+      unsigned int n_vars = sys.n_vars();
+      std::multimap<Real, std::string, std::greater<Real>> ordered_map;
+      for (unsigned int i = 0; i < n_vars; i++)
+      {
+        Real vnorm = sys.calculate_norm(*x, i, DISCRETE_L2);
+        ordered_map.insert(std::pair<Real, std::string>(vnorm, sys.variable_name(i)));
+      }
+
+      std::ostringstream oss;
+      for (const auto & pair : ordered_map)
+        oss << "  {" << pair.second << ", " << pair.first << "},\n";
+
+      mooseError("Aux kernels, user objects appear to have states for aux variables on ",
+                 exec_type,
+                 ".\nVariable error norms in descending order:\n",
+                 oss.str());
+    }
+
+    DenseVector<Real> new_pp_values = getReporterData().getAllRealReporterValues();
+    if (pp_values.size() != new_pp_values.size())
+      mooseError("Second execution for uo/aux state check should not change the number of "
+                 "real reporter values");
+
+    pp_values -= new_pp_values;
+    if (pp_values.l2_norm() > 1e-8)
```

Comment:
I think if you want this to work in parallel as well where there may be different floating point operation orderings we should do more of a relative check instead of absolute